### ISSUE TITLE
Beating PyTorch flash attention in speed

### ISSUE THREAD

**Issue by davnords** [apple/ml-sigmoid-attention#5]
Hi, awesome job, very cool repo!

I was playing around with this seeing if I could get my ViT to run faster using flash sigmoid attention rather than PyTorch's built in flash attention. 

After compiling the kernel I implement the simple code below in my Attention layer and then I simply add torch.amp to make sure it is using the correct precision to work with your code. 

The problem I have is that when I profile this in speed, I get a slower Attention layer than when I use PyTorch's `F.scaled_dot_product_attention`. What am I doing wrong :D? 

This is with B=4, C=1024, N_heads=16 and Nq = 196. 

Thank you very much for your time and effort! 

```python
from flash_sigmoid import flash_attn_func as flash_sigmoid_func

elif self.attn_type == 'flash':
    x = F.scaled_dot_product_attention(
        q, k, v,
        dropout_p=self.attn_drop.p if self.training else 0.,
    )
elif self.attn_type == 'sigmoid':
    x = flash_sigmoid_func(
        q, k, v,
        dropout_p=self.attn_drop.p if self.training else 0.,
    )
```

**Comment by davnords** [comment]
Ahh I think I found the "issue", I tested much longer sequence length (for example N=10,000) and then sigmoid is much faster. Than PyTorch's flash attention

**Comment by jramapuram** [comment]
Hey thanks @davnords : the problem with ViTs is that MLPs typically dominate the computational complexity. For example a 224 x 224 image with a 16 x 16 window => 196 tokens. This  is `O( 196 ^ 2)` for attention which is << `O(embed_dim * embed_dim * 4 *2)` for the MLP.

You should notice a speedup much earlier than 10k tokens though 

<img width="1346" alt="Image" src="https://github.com/user-attachments/assets/7dcd48e2-ff43-4ece-825d-9d0add0045e5" />

**Comment by davnords** [comment]
Thank you @jramapuram! Yes you are right. I did some benchmarking and on images of 512x512 you get some speedup but it really starts being a hyper-speed up at 1024x1024 (if you use patch size 16 like me :)) 

/ David
